### PR TITLE
Bugfix: type not being set when inferring types+shapes

### PR DESCRIPTION
**Pull request by lukeyeager**
/cc @akyrola 

I basically just copied all the `ShapeCall` stuff as `TypeCall`. Is there a better way?

### CODE REVIEW

File: caffe2/core/context_gpu.cu [hunk 1]
```diff
@@ -136,6 +136,11 @@ static void Caffe2InitializeCuda() {
   // Restore the current device.
   CUDA_ENFORCE(cudaSetDevice(init_device));
 
+  RegisterTypeCallFunction(
```
Comment (akyrola) [inline caffe2/core/context_gpu.cu:139]: We need to do the same to Tensor<CPUContext>? I think RegisterShapeCallFunction for TensorCPU is done in tensor.cc

Comment (lukeyeager) [reply caffe2/core/context_gpu.cu:139]: It's in the constructor for the registry

Comment (akyrola) [reply caffe2/core/context_gpu.cu:139]: Oh yeah, sorry!

**Review by akyrola — APPROVED** [review]

**Comment by lukeyeager** [comment]
@Yangqing GitHub reports a merge conflict, but @facebook-github-bot says you already imported it. I'm not sure what it means when you import a PR, exactly. Do you need me to rebase this?

**Comment by Yangqing** [comment]
Thanks @lukeyeager - yeah, if you don't mind I'll just merge it internally, and also resolve some of our internal tests that depends on the test interface change. does that work?

**Comment by lukeyeager** [comment]
Ah got it. Didn't realize I was breaking your stuff. Ok, I'll leave this un-rebased.

**Comment by Yangqing** [comment]
Actually, @lukeyeager , my apologies, I mistook this for #199 - could you rebase?

(that PR is landing in a minute btw)

**Pull request closed without merging**